File: src/main/java/com/itops/itopsagent/service/agent/ContextBuilder.java
package com.itops.itopsagent.service.agent;

import com.itops.itopsagent.dto.AgentContextResponse;
import com.itops.itopsagent.dto.ConversationMessageResponse;
import com.itops.itopsagent.dto.TicketContextResponse;
import com.itops.itopsagent.entity.Ticket;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.springframework.stereotype.Component;

@Component
public class ContextBuilder {

    public AgentContextSnapshot buildSnapshot(Ticket ticket, TicketContextResponse context, List<ConversationMessageResponse> recentMessages) {
        return new AgentContextSnapshot(
                Map.of(
                        "ticketId", ticket.getTicketId(),
                        "title", ticket.getTitle(),
                        "description", ticket.getDescription(),
                        "creatorId", ticket.getCreatorId(),
                        "creatorRole", ticket.getCreatorRole().name(),
                        "status", ticket.getStatus().name()),
                Map.of(
                        "status", ticket.getStatus().name(),
                        "version", ticket.getVersion(),
                        "updatedAt", ticket.getUpdatedAt().toString()),
                context.slots(),
                context.missingSlots(),
                recentMessages.stream().map(message -> Map.<String, Object>of(
                                "id", message.id(),
                                "role", message.role().name(),
                                "content", message.content(),
                                "messageType", message.messageType().name(),
                                "createdAt", message.createdAt().toString()))
                        .toList(),
                "",
                context.matchedSopIds(),
                List.of(),
                Map.of(),
                Map.of(),
                context.lastAgentStep(),
                context.intent());
    }

    public AgentContextResponse buildContextResponse(Ticket ticket, TicketContextResponse context, List<ConversationMessageResponse> recentMessages) {
        AgentContextSnapshot snapshot = buildSnapshot(ticket, context, recentMessages);
        return new AgentContextResponse(
                snapshot.ticketFacts(),
                snapshot.currentState(),
                snapshot.knownSlots(),
                snapshot.missingSlots(),
                snapshot.recentMessages(),
                snapshot.conversationSummary(),
                snapshot.matchedSops(),
                snapshot.toolEvidence(),
                snapshot.approvalContext(),
                snapshot.riskPolicy(),
                snapshot.currentNode());
    }

    public Map<String, Object> buildCompletionSummary(TicketContextResponse context) {
        Map<String, Object> summary = new LinkedHashMap<>();
        summary.put("intent", context.intent().name());
        summary.put("slots", context.slots());
        summary.put("missingSlots", context.missingSlots());
        summary.put("lastAgentStep", context.lastAgentStep());
        return summary;
    }
}

File: agent_runtime/context/context_builder.py
"""按当前节点组装最小上下文，不把全量历史直接塞给模型。"""

from __future__ import annotations


class ContextBuilder:
    def build(self, ticket: dict, ticket_context: dict | None, recent_messages: list[dict]) -> dict:
        ticket_context = ticket_context or {}
        return {
            "ticket_facts": {
                "ticketId": ticket["ticketId"],
                "title": ticket["title"],
                "description": ticket["description"],
                "status": ticket["status"],
            },
            "current_state": {
                "status": ticket["status"],
                "version": ticket["version"],
            },
            "known_slots": ticket_context.get("slots", {}),
            "missing_slots": ticket_context.get("missingSlots", []),
            "recent_messages": recent_messages[-6:],
            "conversation_summary": ticket_context.get("conversationSummary", ""),
            "matched_sops": ticket_context.get("matchedSops", []),
            "tool_evidence": [],
            "approval_context": {},
            "risk_policy": {},
            "current_node": ticket_context.get("lastAgentStep", "INIT"),
        }
